src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -120,11 +125,20 @@ protected void ShutdownAllNodes(bool nodeReuse, NodeContextTerminateDelegate ter
             // Search for all instances of msbuildtaskhost process and add them to the process list
             nodeProcesses.AddRange(new List<Process>(Process.GetProcessesByName(Path.GetFileNameWithoutExtension(msbuildtaskhostExeName))));
 
+            ShutDownNodes(nodeProcesses, nodeReuse, terminateNode);
+        }
+
+        private void ShutDownNodes(List<Process> nodeProcesses, bool nodeReuse, NodeContextTerminateDelegate terminateNode)
+        {
+            // INodePacketFactory
+            INodePacketFactory factory = new NodePacketFactory();
+
             // For all processes in the list, send signal to terminate if able to connect
             foreach (Process nodeProcess in nodeProcesses)
             {
                 // A 2013 comment suggested some nodes take this long to respond, so a smaller timeout would miss nodes.
-                int timeout = 30;
+                // Updated 2021 to correspond to new times taken to respond.
```

Comment:
How did you measure this? Could we add the time actually taken here to the ETW-to-text logging, then ask folks to turn that on if they ever see this problem and dump it if they see it again?

Reply:
Two people have collected comm logs (with fewer MSBuild nodes running, though still an excess), and in looking at one of them, it seemed that the timeout was just a little too short for shutting down nodes (up to ~43 ms). ETW trace could be nice.